### PR TITLE
Issue 878

### PR DESCRIPTION
**Pull request by kruger**
This is following the outline in the discussion of Issue 878.  The commits are left individually to see the process and see if there is an implementation path that is disliked.  Perhaps the most controversial part is that I changed how `KokkosCore_config.h` is generated, and the switching from a `$(shell )` design pattern to a `$(call )` pattern.   The first two commits are the part that should be viewed most carefully.

### CODE REVIEW

**Comment by crtrott** [comment]
What are all those unit_test/config/... files? Were those just accidentally added?

**Comment by crtrott** [comment]
General comment: when designing the Make based system we opted generally for simple features which can be understood with basic Make knowledge. It seems to me that a lot of the changes go away from that, with the added question of how portable those new used features are (i.e. how reliable do they work across all kinds of different OS (versions).

**Comment by crtrott** [comment]
Another general question: all Makefile* in all of Kokkos (including all the examples tutorials etc.) are 4k lines. This commit adds >8k is that really necessary or is a big part of that accidental commit of files (see question above).

File: CMakeLists.txt [hunk 1]
```diff
@@ -4,11 +4,27 @@ ELSE()
   SET(KOKKOS_HAS_TRILINOS OFF CACHE BOOL "")
 ENDIF()
 
+set(Kokkos_SOURCE_DIR ${CMAKE_SOURCE_DIR})
+
 IF(NOT KOKKOS_HAS_TRILINOS)
   cmake_minimum_required(VERSION 3.1 FATAL_ERROR)
   project(Kokkos CXX)
 
-  INCLUDE(cmake/kokkos.cmake)
```
Comment (ibaned) [inline CMakeLists.txt]: Does this PR preserve the behavior of this "pure CMake" build system that @bjoo and others rely on?

File: CMakeLists.txt [hunk 2]
```diff
@@ -4,11 +4,27 @@ ELSE()
...
+  # Handle the options
+  include(${CMAKE_SOURCE_DIR}/cmake/kokkos_options.cmake)
+
+  #---------------------------------------------------------------------
+  # File names must be consistent with core/src/Makefile
+  # This generates the ${CMAKE_BINARY_DIR}/gen_kokkos.cmake command
```
Comment (ibaned) [inline CMakeLists.txt]: The last word in the comment should be "file", not "command"

File: Makefile.kokkos [hunk 3]
```diff
@@ -24,33 +24,36 @@ KOKKOS_OPTIONS ?= ""
 # Options: force_uvm,use_ldg,rdc,enable_lambda
 KOKKOS_CUDA_OPTIONS ?= ""
 
+#TODO - I think this will work:
+ifhas=$(if $(findstring $2,$1),1,0)
```
Comment (ibaned) [inline Makefile.kokkos]: is it really necessary to replace how this is done? I know @crtrott had concerns, and the comment above this line is not encouraging.

File: Makefile.kokkos [hunk 4]
```diff
@@ -258,43 +264,43 @@ ifeq ($(KOKKOS_INTERNAL_USE_ARCH_NVIDIA), 1)
...
+KOKKOS_INTERNAL_USE_ARCH_VEGA := $(call ifhas,$(KOKKOS_ARCH),Vega)
+KOKKOS_INTERNAL_USE_ARCH_GFX901 := $(call ifhas,$(KOKKOS_ARCH),gfx901)
 
 # Any AVX?
-KOKKOS_INTERNAL_USE_ARCH_SSE42      := $(strip $(shell echo $(KOKKOS_INTERNAL_USE_ARCH_WSM) | bc ))
-KOKKOS_INTERNAL_USE_ARCH_AVX        := $(strip $(shell echo $(KOKKOS_INTERNAL_USE_ARCH_SNB)+$(KOKKOS_INTERNAL_USE_ARCH_AMDAVX) | bc ))
-KOKKOS_INTERNAL_USE_ARCH_AVX2       := $(strip $(shell echo $(KOKKOS_INTERNAL_USE_ARCH_HSW)+$(KOKKOS_INTERNAL_USE_ARCH_BDW) | bc ))
-KOKKOS_INTERNAL_USE_ARCH_AVX512MIC  := $(strip $(shell echo $(KOKKOS_INTERNAL_USE_ARCH_KNL) | bc ))
-KOKKOS_INTERNAL_USE_ARCH_AVX512XEON := $(strip $(shell echo $(KOKKOS_INTERNAL_USE_ARCH_SKX) | bc ))
+KOKKOS_INTERNAL_USE_ARCH_SSE42      := $(shell expr $(KOKKOS_INTERNAL_USE_ARCH_WSM))
```
Comment (ibaned) [inline Makefile.kokkos]: replacing `bc` with `expr` is more reasonable than the above, but again this seems like something that should be a separate PR.

File: Makefile.kokkos [hunk 5]
```diff
@@ -321,89 +327,106 @@ KOKKOS_LDFLAGS = -L$(shell pwd)
...
+	  @$(call wrtkks,$(shell date))
+	  @$(call wrtkks,"----------------------------------------------*/")
+	  
+	  @$(call wrtkks,'#if !defined(KOKKOS_MACROS_HPP) || defined(KOKKOS_CORE_CONFIG_H)')
+	  @$(call wrtkks,'#error "Do not include $(gen_header) directly; include Kokkos_Macros.hpp instead."')
+	  @$(call wrtkks,'#else')
+	  @$(call wrtkks,'#define KOKKOS_CORE_CONFIG_H')
+	  @$(call wrtkks,'#endif')
+	  
+defexes="/* Execution Spaces */"
```
Comment (ibaned) [inline Makefile.kokkos]: again, is it really necessary to have a new system for doing this? and in the same PR?

File: CMakeLists.txt (hunk 1 above)
Comment (kruger) [reply CMakeLists.txt]: In theory yes, in practice, it needs more testing to ensure that it is working.  The last commit shows the changes.

File: Makefile.kokkos (hunk 3 above)
Comment (kruger) [reply Makefile.kokkos]: Sorry -- crufty comment.  It works.  Related to @crtrott comment above:  The `$(call ...)` functionality is supported everywhere other `$(...)` functionality is supported.  The use of `$(shell ...)` and `$(wildcard ...)` gave me confidence that I could use it.

In my personal testing, when I had a build going on my machine and bogging it down, the using of the internal shell functionality was noticably faster (removing ~80 calls out to a shell).

Comment (ibaned) [reply Makefile.kokkos]: > noticably faster (removing ~80 calls out to a shell)

okay, I guess this is a good reason to try to change this

File: Makefile.kokkos (hunk 5 above)
Comment (kruger) [reply Makefile.kokkos]: I view this as the most important part of the initial refactor.  Previously these were done with a `echo ... >> Makefile.kokkos` in one section and an `echo ... > kokkos.cmake` in another section.  This then requires a developer to ensure that variables are provided in two separate places.  By using these functions, we only have a single reducing the chances for developer error.  Since the overall goal is to reduce the build system redundancy, this seemed in keeping with that redundancy.

File: CMakeLists.txt (hunk 2 above)
Comment (kruger) [reply CMakeLists.txt]: Yes.  Thanks.

**Comment by mhoemmen** [comment]
Are y'all assuming GNU Make btw?  Some of those calls look rather GNU Make - specific, but perhaps that's OK.

File: Makefile.kokkos (hunk 5 above)
Comment (ibaned) [reply Makefile.kokkos]: > requires a developer to ensure that variables are provided in two separate places. By using these functions, we only have a single reducing the chances for developer error.

okay, that seems like good rationale too

**Comment by kruger** [comment]
Let me first address the "unit_test".

The files in `unit_test` were not accidental but rather how I developed a unit test for the configuration step itself.  If you `cd core/unit_test; make -f UnitTest` then you can see the results.  

It is doing all of the combinations of arch and devices and then comparing them with "accepted results".  I understand that this method of testing against accepted results is not a typical unit_test design method.   My motivation here is that the logic in `$(KOKKOS_SRC_PATH)/Makefile.kokkos` and again in `$(KOKKOS_SRC_PATH)/core/src/Makefile` is the valuable part of the build system as it has the routed if statements and logic on how to respond to the architectures and devices, and that having the entire routing tested on a single device would be valuable.

The possibilities here are:

1. Unit testing the build system is not worth having in the repo.   _Response: Fine, I can remove it_

2.  Unit testing the build system is a good idea, but the implementation is bad. _Response: Fine, but I am unsure of a good way of testing this in any other way.

**Comment by kruger** [comment]
Regarding the number of lines:   It is pretty much all unit testing.  Looking at key files:

- `Makefile.kokkos`: ~35 extra lines mostly due to changing the generation of `KokkosCore_config.h` to being generated by targets.

- `core/src/Makefile`: Approximately the same number of lines

-`cmake/kokkos.cmake`:  Split up into 3 separate files, but ~350 fewer lines from replacing the logic with the Makefile logic.

**Comment by ibaned** [comment]
> It is doing all of the combinations of arch and devices and then comparing them with "accepted results"

Okay, for me it was just unclear what their purpose was at first. That seems reasonable to me then, and I can't quite think of a much better way than having "gold" files.

**Comment by kruger** [comment]
@mhoemmen   The `$(call ...)` syntax is equivalent to using the `$(wildcard ...)`,  `$(shell ...)`, `$(strip ..)`  that was already being used.  The `ifhas` function uses also uses `$(findstring ...)` but that is also previously used in the example makefiles.

**Comment by kruger** [comment]
General comment:  I have tested the Makefile builds extensively and feel good about them.  The cmake builds do not have tests and right now I can only ensure that I went through a build with them.  Understanding how they are tested would be useful.  

In comparing develop and this branch, I do see differences in the files that are compiled into the library with the cmake builds.  This is because `cmake/kokkos.cmake` uses a different method for calculating the sources from `Makefile.kokkos`.  Because I trust the logic in `Makefile.kokkos` more, I think this is OK, but testing is really needed.

**Comment by dholladay00** [comment]
`/example/cmake_build` contains how standalone cmake is used. I use it in this way. @bjoo, anything to add?

**Comment by kruger** [comment]
@dholladay00  Thanks -- I'll take a look.

**Comment by kruger** [comment]
I am trying to reconcile some of the differences between the CMake builds and the Makefile.  As an example of what I am trying to understand, try `git  grep CUDA_RDC` or `git grep CUDA_LAMBDA`.

The fact that `<KOKKOS_PATH>/cmake/kokkos.cmake` and `<KOKKOS_PATH>/core/cmake/KokkosCore_config.h.in` have these variables is mostly legacy, correct?   In other words, the Makefile system isn't doing anything to put these variables into `KokkosCore_config.h` with `Makefile.kokkos` because they just are no longer needed.  Correct?

**Comment by crtrott** [comment]
Actually its a stupid mixup. In one case the Makefile system is correct in the other it is not . 
The correct macros should be:
```
KOKKOS_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE
KOKKOS_ENABLE_CUDA_LAMBDA
```

Old versions are:
```
KOKKOS_HAVE_CUDA_RDC
KOKKOS_HAVE_CUDA_LAMBDA
KOKKOS_CUDA_USE_LAMBDA
```

File: Makefile.kokkos [hunk 6]
```diff
@@ -24,33 +24,36 @@ KOKKOS_OPTIONS ?= ""
 # Options: force_uvm,use_ldg,rdc,enable_lambda
 KOKKOS_CUDA_OPTIONS ?= ""
 
+#TODO - I think this will work:
```
Comment (crtrott) [inline Makefile.kokkos]: Okay, we need some comment here and also I'd like to have a systematic naming of functions which get added and are not native to cmake. Lets for example do the following convention:

Variable all start with KOKKOS_, variables not for consumation by application Makefiles start with KOKKOS_INTERNAL.

Functions start with kokkos_ (note lower case) so this ifhas function would become kokkos_ifhas.

File: Makefile.kokkos (hunk 3 above)
Comment (crtrott) [reply Makefile.kokkos]: I'd propose a namechange of the function to 'kokkos_has_string'

File: Makefile.kokkos [hunk 7]
```diff
@@ -321,89 +327,106 @@ KOKKOS_LDFLAGS = -L$(shell pwd)
 KOKKOS_SRC =
 KOKKOS_HEADERS =
 
-# Generating the KokkosCore_config.h file.
 
-tmp := $(shell echo "/* ---------------------------------------------" > KokkosCore_config.tmp)
-tmp := $(shell echo "Makefile constructed configuration:" >> KokkosCore_config.tmp)
-tmp := $(shell date >> KokkosCore_config.tmp)
-tmp := $(shell echo "----------------------------------------------*/" >> KokkosCore_config.tmp)
+gen_config=KokkosCore_config.tmp
```
[the remark this answers is not in the document]
Comment (crtrott) [reply Makefile.kokkos]: Proposed name: KOKKOS_INTERNAL_CONFIG_TMP

File: Makefile.kokkos [hunk 8]
```diff
@@ -321,89 +327,106 @@ KOKKOS_LDFLAGS = -L$(shell pwd)
...
+gen_header=KokkosCore_config.h
```
Comment (crtrott) [inline Makefile.kokkos]: Proposed name: KOKKOS_CONFIG_HEADER

Comment (crtrott) [reply Makefile.kokkos]: Reason for its not KOKKOS_INTERNAL: other folks may make something directly dependent on this file.

File: Makefile.kokkos [hunk 9]
```diff
@@ -321,89 +327,106 @@ KOKKOS_LDFLAGS = -L$(shell pwd)
...
+# Functions for generating config header file
+wrtkks = echo $1 >> $(gen_config)
```
Comment (crtrott) [inline Makefile.kokkos]: Rename kokkos_append_header

File: Makefile.kokkos (hunk 5 above)
Comment (crtrott) [reply Makefile.kokkos]: I am starting to be ok with this, but check my comments above for naming changes (I will summarize those later in a comment).

Comment (crtrott) [reply Makefile.kokkos]: Can you elaborate on why this need to happen in targets instead of the direct append to files while the Makefile is parsed? It seems to add some complexity for now without apparent reason?

File: Makefile.kokkos [hunk 10]
```diff
@@ -321,89 +327,106 @@ KOKKOS_LDFLAGS = -L$(shell pwd)
...
-tmp := $(shell echo '\#if !defined(KOKKOS_MACROS_HPP) || defined(KOKKOS_CORE_CONFIG_H)' >> KokkosCore_config.tmp)
-tmp := $(shell echo '\#error "Do not include KokkosCore_config.h directly; include Kokkos_Macros.hpp instead."' >> KokkosCore_config.tmp)
-tmp := $(shell echo '\#else' >> KokkosCore_config.tmp)
-tmp := $(shell echo '\#define KOKKOS_CORE_CONFIG_H' >> KokkosCore_config.tmp)
-tmp := $(shell echo '\#endif' >> KokkosCore_config.tmp)
+# Generating the KokkosCore_config.h file.
 
-tmp := $(shell echo "/* Execution Spaces */" >> KokkosCore_config.tmp)
+defines=""
+define1=""
```
Comment (crtrott) [inline Makefile.kokkos]: Rename: KOKKOS_INTERNAL_CONFIG_DEFINES and KOKKOS_INTERNAL_CONFIG_DEFINES_1

File: Makefile.kokkos (hunk 5 above)
Comment (crtrott) [reply Makefile.kokkos]: Rename: defexes -> KOKKOS_INTERNAL_CONFIG_DEFINES_EXEC

**Comment by crtrott** [comment]
There are quite a few defines and function names which need changes: 
```
ifhas -> kokkos_has_string
gen_config -> KOKKOS_INTERNAL_CONFIG_TMP
gen_header -> KOKKOS_CONFIG_HEADER
wrtkks -> kokkos_append_header
defines -> KOKKOS_INTERNAL_CONFIG_DEFINES
defines1 -> KOKKOS_INTERNAL_CONFIG_DEFINES_1
defexes -> KOKKOS_INTERNAL_CONFIG_DEFINES_EXEC
def... -> KOKKOS_INTERNAL_CONFIG_DEFINES_...
```

**Comment by crtrott** [comment]
Scott can you maybe phone in for our weekly Kokkos meeting tomorrow at noon mountain time?

**Comment by crtrott** [comment]
Scott can you maybe phone in for our weekly Kokkos meeting tomorrow at noon mountain time?

File: Makefile.kokkos (hunk 6 above)
Comment (kruger) [reply Makefile.kokkos]: I'm confused by this comment.  Are you referring to `KOKKOS_CUDA_OPTIONS`?

**Comment by kruger** [comment]
@crtrott Will do if I can have the coordinates.  Working on name changes now.

File: Makefile.kokkos (hunk 6 above)
Comment (crtrott) [reply Makefile.kokkos]: No I mean we need a bit more comment on what the function ifhas does.

**Comment by crtrott** [comment]
I send you an email with the call info did you get that?

**Comment by kruger** [comment]
I have now tested on Mac and Linux with a pthread backend and both serial and parallel, and with the 3 different build systems (make, cmake, TriBITS).  I think things are in good shape but more challenging build systems would be good.  This should be coordinated with TriBITS branch kokkos-config-207 and Trilinos branch trilinos-kokkos-1400.

**Comment by bartlettroscoe** [comment]
@kruger,

@crtrott says that we should test this on ellis or white and boman or shiller at Trilinos.  

Let's talk.

**Comment by crtrott** [comment]
I will test this on the ATDM machines, and then merge that in after the current promotion is through. I may run another promotion really soon since I got to do one for KokkosKernels. Might be worthwhile to mainly have this build system overhaul in that one.

**Comment by bartlettroscoe** [comment]
> I will test this on the ATDM machines, and then merge that in after the current promotion is through. I may run another promotion really soon since I got to do one for KokkosKernels. Might be worthwhile to mainly have this build system overhaul in that one.

@crtrott,

We would like to test this with Trilinos as well (which was the motivation for this work in the first place, see trilinos/Trilinos#1400).

Can you give us a short list of of the more challenging ATDM machines where we should try this and what Makefile options (i.e. arch and backend vars you use on these machines)?

**Comment by bartlettroscoe** [comment]
@kruger and I got on and went over the list of ATDM test bed machines and I think we have the right list of machines to start testing on based on what @crtrott told me at the TUG in person last week.  Hopefully we can guess the right arch and backend options to select just from the specs the machines (e.g. we can pretend we are users who don't know anything other than this stuff).

**Comment by kruger** [comment]
On the appropriate open test machine, I have 6 builds with the different permutations in `~sekruge/builds/`.  In each build directory, see the `config.sh` for how it was configured.  It has HSW arch and Cuda, Pthread devices

- par-branches-gcc4.9-1
     o TriBITS build with the new branches (kokkos: issue-878; trilinos: trilinos-kokkos-1400, TriBITS: kokkos-config-207) 

- par-trunk-gcc4.9-1
     o  TriBITS build with the snapshot version of TriBITS and kokkos

- par-kokkos-gcc4.9-1
     o  CMake build of kokkos: issue-878

- par-kokkos-trunkcmake-gcc4.9-1
     o CMake build of snapshot version of kokkos

- par-kokkosmake-gcc4.9-1
    o generate_makefile.bash/makefile build of kokkos: issue-878

- par-kokkos-trunkmake-gcc4.9-1
    o generate_makefile.bash/makefile build of kokkos snapshot

They are all building but I am getting an error when I do a `make build-test` for the makefile builds:
```
...
cp -u   /ascldap/users/sekruge/builds/par-kokkos-trunkmake-gcc4.9-1/install
cp: missing destination file operand after `/ascldap/users/sekruge/builds/par-kokkos-trunkmake-gcc4.9-1/install
```

This is for both the snapshot and issue-878 versions.

I did find one bug with the cmake builds and that is fixed.

So, things are building fine for all of the permutations and the builds seem fine to the extent that I can test.  Advice on how to test further would be welcome.

**Comment by crtrott** [comment]
I will do now some testing (I will also fix the merge conflict) and then proceed to merge this in.

**Comment by crtrott** [comment]
Found the error: there was a not yet renamed gen_makefile in the copy command, which should be MAKEFILE_KOKKOS. But if all the tests stopped there nothing actually got build except the library. I am going to do a comprehensive test now.

**Comment by crtrott** [comment]
Starting a test matrix tracking here (will modify the post to update):

| Machine | Compiler | BuildSystem | Backends | Options | Status |
|---|---|---|---|---|---|
| white | gcc 5.4 | make | OpenMP | - | Passed |

**Comment by crtrott** [comment]
Ok I believe our plain cmake build is broken with this patch. It neither adds the architecture flags nor the openmp flag when I do this:

cmake -DKOKKOS_HOST_ARCH="SNB" /home/crtrott/Kokkos/kokkos

It does with develop.

**Comment by crtrott** [comment]
Found the issue for the above problem: mismatch between KOKKOS_CXX_FLAGS and KOKKOS_CXXFLAGS. 
Now trying to find out why -Xcompiler is appended ...

**Comment by crtrott** [comment]
OK I fixed some issues. Now going on to more testing.

**Comment by crtrott** [comment]
Yet another issue when trying to install:

```
CMake Error at cmake_install.cmake:72 (file):
  file INSTALL cannot find "/ascldap/users/crtrott/Kokkos/kokkos/EXPORT".
```

**Comment by crtrott** [comment]
Ok here are two configures which work on my desktop:
```
cmake -DKOKKOS_ARCH="SNB;Kepler35" -DCMAKE_CXX_COMPILER=/home/crtrott/Kokkos/kokkos/bin/nvcc_wrapper -DKOKKOS_ENABLE_CUDA=True -DCMAKE_INSTALL_PREFIX="/home/crtrott/Kokkos/KokkosBuild/OpenMP-CMake2/install" -DKokkos_ENABLE_TESTS=on -DCMAKE_CXX_EXTENSIONS=OFF /home/crtrott/Kokkos/kokkos
```
```
cmake -DKOKKOS_ARCH="SNB" -DKOKKOS_ENABLE_OPENMP=ON -DCMAKE_INSTALL_PREFIX="/home/crtrott/Kokkos/KokkosBuild/OpenMP-CMake/install" -DKokkos_ENABLE_TESTS=on /home/crtrott/Kokkos/kokkos
```
